File: Game-1-modular/data/models/skill_unlocks.py
"""Skill unlock data models"""

from dataclasses import dataclass, field
from typing import Dict, List, Any, Optional

from data.models.unlock_conditions import UnlockRequirements


@dataclass
class UnlockCost:
    """Cost to unlock a skill after conditions are met."""
    gold: int = 0
    materials: List[Dict[str, Any]] = field(default_factory=list)  # [{materialId, quantity}]
    skill_points: int = 0

    def can_afford(self, character) -> tuple[bool, str]:
        """
        Check if character can afford the unlock cost.

        Args:
            character: Character to check

        Returns:
            (can_afford, reason): True if affordable, reason if not
        """
        # Check gold
        if hasattr(character, 'gold') and character.gold < self.gold:
            return False, f"Need {self.gold} gold"

        # Check materials
        for mat_req in self.materials:
            mat_id = mat_req.get('materialId')
            qty = mat_req.get('quantity', 1)
            if not character.inventory.has_item(mat_id, qty):
                return False, f"Need {qty}x {mat_id}"

        # Check skill points (if system uses them)
        if self.skill_points > 0:
            if hasattr(character, 'skill_points') and character.skill_points < self.skill_points:
                return False, f"Need {self.skill_points} skill points"

        return True, "OK"

    def pay(self, character) -> bool:
        """
        Pay the unlock cost.

        Args:
            character: Character paying the cost

        Returns:
            True if successfully paid
        """
        can_afford, reason = self.can_afford(character)
        if not can_afford:
            return False

        # Deduct gold
        if hasattr(character, 'gold'):
            character.gold -= self.gold

        # Remove materials
        for mat_req in self.materials:
            mat_id = mat_req.get('materialId')
            qty = mat_req.get('quantity', 1)
            character.inventory.remove_item(mat_id, qty)

        # Deduct skill points
        if self.skill_points > 0 and hasattr(character, 'skill_points'):
            character.skill_points -= self.skill_points

        return True


@dataclass
class UnlockTrigger:
    """Defines when/how a skill unlock actually happens."""
    type: str  # level_up, quest_complete, title_earned, activity_threshold, etc.
    trigger_value: Any  # Context-specific (questId, level, count, etc.)
    message: str  # Notification shown to player


@dataclass
class SkillUnlock:
    """
    Defines how a skill becomes available to the player.

    Uses tag-driven UnlockRequirements for conditions.
    """
    unlock_id: str
    skill_id: str
    unlock_method: str  # automatic, quest_reward, milestone_unlock, title_unlock, etc.

    requirements: UnlockRequirements
    trigger: UnlockTrigger
    cost: UnlockCost

    # Metadata
    narrative: str = ""
    category: str = ""

    def __post_init__(self):
        """Validate unlock definition."""
        if not self.unlock_id:
            raise ValueError("unlock_id is required")
        if not self.skill_id:
            raise ValueError("skill_id is required")
        if not self.unlock_method:
            raise ValueError("unlock_method is required")

    def check_conditions(self, character) -> bool:
        """
        Check if unlock conditions are satisfied.

        Args:
            character: Character to check

        Returns:
            True if all conditions met
        """
        return self.requirements.evaluate(character)

    def check_cost(self, character) -> tuple[bool, str]:
        """
        Check if character can afford unlock cost.

        Args:
            character: Character to check

        Returns:
            (can_afford, reason)
        """
        return self.cost.can_afford(character)

    def can_unlock(self, character) -> tuple[bool, str]:
        """
        Check if skill can be unlocked (conditions + cost).

        Args:
            character: Character to check

        Returns:
            (can_unlock, reason)
        """
        if not self.check_conditions(character):
            return False, "Conditions not met"

        can_afford, reason = self.check_cost(character)
        if not can_afford:
            return False, reason

        return True, "OK"

    def unlock(self, character) -> tuple[bool, str]:
        """
        Unlock the skill for the character.

        Args:
            character: Character unlocking the skill

        Returns:
            (success, message)
        """
        can_unlock, reason = self.can_unlock(character)
        if not can_unlock:
            return False, reason

        # Pay cost
        if not self.cost.pay(character):
            return False, "Failed to pay cost"

        # Add skill to character's skill manager
        if hasattr(character, 'skills'):
            # Mark skill as unlocked in skill manager
            # The skill manager will handle actually adding the skill
            success = character.skills.unlock_skill(self.skill_id)
            if success:
                return True, self.trigger.message
            else:
                return False, f"Failed to unlock skill {self.skill_id}"

        return False, "Character has no skill manager"

File: Unity/Assets/Scripts/Game1.Data/Models/SkillUnlocks.cs
// Game1.Data.Models.SkillUnlocks
// Migrated from: data/models/skill_unlocks.py (182 lines)
// Phase: 1 - Foundation
// Contains: UnlockCost, UnlockTrigger, SkillUnlock
// Note: Methods referencing Character use object parameter, deferred to Phase 3.

using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Game1.Data.Models
{
    /// <summary>
    /// Cost to unlock a skill after conditions are met.
    /// </summary>
    [Serializable]
    public class UnlockCost
    {
        [JsonProperty("gold")]
        public int Gold { get; set; }

        [JsonProperty("materials")]
        public List<Dictionary<string, object>> Materials { get; set; } = new();

        [JsonProperty("skillPoints")]
        public int SkillPoints { get; set; }

        /// <summary>
        /// Check if character can afford the unlock cost.
        /// Uses object parameter; concrete Character checking deferred to Phase 3.
        /// </summary>
        public (bool CanAfford, string Reason) CanAfford(object character)
        {
            // Phase 3 will implement concrete Character-based checks:
            // - Gold check: character.Gold >= Gold
            // - Material check: character.Inventory.HasItem(matId, qty)
            // - Skill point check: character.SkillPoints >= SkillPoints
            return (true, "OK");
        }

        /// <summary>
        /// Pay the unlock cost. Returns true if successfully paid.
        /// Concrete implementation deferred to Phase 3.
        /// </summary>
        public bool Pay(object character)
        {
            var (canAfford, _) = CanAfford(character);
            if (!canAfford) return false;
            // Phase 3 will implement actual deduction
            return true;
        }
    }

    /// <summary>
    /// Defines when/how a skill unlock actually happens.
    /// </summary>
    [Serializable]
    public class UnlockTrigger
    {
        [JsonProperty("type")]
        public string Type { get; set; }

        [JsonProperty("triggerValue")]
        public object TriggerValue { get; set; }

        [JsonProperty("message")]
        public string Message { get; set; }
    }

    /// <summary>
    /// Defines how a skill becomes available to the player.
    /// Uses tag-driven UnlockRequirements for conditions.
    /// </summary>
    [Serializable]
    public class SkillUnlock
    {
        [JsonProperty("unlockId")]
        public string UnlockId { get; set; }

        [JsonProperty("skillId")]
        public string SkillId { get; set; }

        [JsonProperty("unlockMethod")]
        public string UnlockMethod { get; set; }

        [JsonProperty("requirements")]
        public UnlockRequirements Requirements { get; set; }

        [JsonProperty("trigger")]
        public UnlockTrigger Trigger { get; set; }

        [JsonProperty("cost")]
        public UnlockCost Cost { get; set; }

        [JsonProperty("narrative")]
        public string Narrative { get; set; } = "";

        [JsonProperty("category")]
        public string Category { get; set; } = "";

        /// <summary>
        /// Validate unlock definition. Throws if required fields are missing.
        /// </summary>
        public void Validate()
        {
            if (string.IsNullOrEmpty(UnlockId))
                throw new ArgumentException("unlockId is required");
            if (string.IsNullOrEmpty(SkillId))
                throw new ArgumentException("skillId is required");
            if (string.IsNullOrEmpty(UnlockMethod))
                throw new ArgumentException("unlockMethod is required");
        }

        public bool CheckConditions(object character) =>
            Requirements?.Evaluate(character) ?? true;

        public (bool CanAfford, string Reason) CheckCost(object character) =>
            Cost?.CanAfford(character) ?? (true, "OK");

        public (bool CanUnlock, string Reason) CanUnlock(object character)
        {
            if (!CheckConditions(character))
                return (false, "Conditions not met");

            var (canAfford, reason) = CheckCost(character);
            if (!canAfford) return (false, reason);

            return (true, "OK");
        }

        /// <summary>
        /// Unlock the skill for the character. Concrete implementation in Phase 3.
        /// </summary>
        public (bool Success, string Message) Unlock(object character)
        {
            var (canUnlock, reason) = CanUnlock(character);
            if (!canUnlock) return (false, reason);

            if (Cost != null && !Cost.Pay(character))
                return (false, "Failed to pay cost");

            // Phase 3 will add skill to character's SkillManager
            return (true, Trigger?.Message ?? "Skill unlocked!");
        }
    }
}
